### PR TITLE
reactor: add process::terminate() and process::kill()

### PR DESCRIPTION
**Pull request by tchaikov**
so we can terminiate a subprocess

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
this helper is added, so we can fork an HTTPS server as the fixture for testing TLS testings like:

```c++
template <typename Func>
auto with_https_server(Func&& func) {
    auto httpd = boost::dll::program_location().parent_path() / "https-server.py";
    const std::vector<sstring> argv{
      "httpd",
      "--server", "0.0.0.0:11443",
      "--key", certfile("mtls_server.key"),
      "--cert", certfile("mtls_server.crt"),
    };
    return experimental::spawn_process(httpd.string(), {.argv = argv}).then([func = std::move(func)] (auto process) {
        return do_with(std::move(process), [func = std::move(func)](auto& p) {
            return func().finally([&p] {
                return p.terminate().then([&p] {
                    return p.wait().discard_result();
                });
            });
        });
    });
}

/// 

SEASTAR_TEST_CASE(yet_another_tls_test) {
    return with_https_server([] {
        return connect_to_https_server_with_tls();
    });
}
```

**Comment by tchaikov** [comment]
hello? ping for reviews.

File: include/seastar/util/process.hh [hunk 1]
```diff
@@ -87,6 +87,11 @@ public:
     ///
     /// \returns the exit status
     future<wait_status> wait();
+    /// Stop the process using SIGTERM
+    future<> terminate();
+    /// Force the process to exit using SIGKILL
+    future<> kill();
```
Comment (avikivity) [inline include/seastar/util/process.hh:93]: Do those need to return futures?

File: src/core/reactor.cc [hunk 2]
```diff
@@ -2061,6 +2061,15 @@ future<int> reactor::waitpid(pid_t pid) {
     });
 }
 
+future<> reactor::kill(pid_t pid, int sig) {
+    return _thread_pool->submit<syscall_result<int>>([pid, sig] {
+        return wrap_syscall<int>(::kill(pid, sig));
+    }).then([] (syscall_result<int> ret) {
+        ret.throw_if_error();
+        return make_ready_future();
+    });
```
Comment (avikivity) [inline src/core/reactor.cc:2070]: Why use the threadpool? kill() should be non-blocking.

File: include/seastar/util/process.hh (hunk 1 above)
Comment (tchaikov) [reply include/seastar/util/process.hh]: hi @avikivity ahh, right. thought it was syscall, and syscall could be slow in comparison to user space function calls. but yeah, `kill()` does not block. so updated accordingly.

**Comment by tchaikov** [comment]
changelog

- do not use thread pool for calling `kill(2)`.
- do not let `kill()` return `future<>`

File: tests/unit/spawn_test.cc [hunk 3]
```diff
@@ -116,3 +116,23 @@ SEASTAR_TEST_CASE(test_spawn_input) {
...
+    return spawn_process(sleep_cmd, {.argv = {sleep_cmd, "10"}}).then([] (auto process) {
+        auto start = std::chrono::high_resolution_clock::now();
+        return do_with(std::move(process), [](auto& p) {
+            p.terminate();
+            return p.wait();
+        }).then([start](experimental::process::wait_status wait_status) {
+            auto* wait_signaled = std::get_if<experimental::process::wait_signaled>(&wait_status);
+            BOOST_REQUIRE(wait_signaled != nullptr);
+            BOOST_CHECK_EQUAL(wait_signaled->terminating_signal, SIGTERM);
+            // sleep should be terminated in 10ms
```
Comment (balusch) [inline tests/unit/spawn_test.cc:132]: How is the number calculated or it's just an estimation?

Comment (tchaikov) [reply tests/unit/spawn_test.cc:132]: it's a rough estimation. 10ms is more than enough for the RTT of local system without involving networking or disk i/o.

Comment (balusch) [reply tests/unit/spawn_test.cc:132]: It takes nearly 20ms on my machine(ArchLinux, running in docker), so weird.

> /path/to/spawn_test.cc(135): error: in "test_spawn_kill": check ms <= 10 has failed [20 > 10]

Comment (tchaikov) [reply tests/unit/spawn_test.cc:132]: 20ms is indeed unexpected. what's the latency between a `kill(2)` call and the completion of `waitpid()` ? i created a tiny tool, see https://godbolt.org/z/rdraqsaMG

and my output is like
```console
$  g++ test_fork.cc
$ ./a.out sleep 10
child killed by signal 15
duration: 000:000:275165
sleep returned -1
```

Comment (balusch) [reply tests/unit/spawn_test.cc:132]: looks quite normal:

```txt
% clang++ kill.cc
% ./a.out sleep 10
child killed by signal 15
duration: 000:000:102323
sleep returned -1
```

Comment (tchaikov) [reply tests/unit/spawn_test.cc:132]: please feel free to file an issue. or if you have a theory explaining this, a PR would be better!

i don't have an explanation for the discrepancy between 102323ns and 20ms. reactor's period of polling should not incur such a huge latency. IMHO, both linux and `sleep` should be very responsive under normal load.

Comment (balusch) [reply tests/unit/spawn_test.cc:132]: sure. I'll make a deep investigation this weekend.

**Pull request merged**